### PR TITLE
New common script to create API review request from artifact

### PR DESCRIPTION
**Pull request by praveenkuttappan**
Common script that will be consumed by each language pipeline to create API review

### CODE REVIEW

File: eng/common/scripts/Create-APIReview.ps1 [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    catch{
+        $StatusCode = $_.Exception.Response.StatusCode
+    }
+
+    return $StatusCode
+}
+
+
+. "${PSScriptRoot}\..\common\scripts\common.ps1"
+$packages = Find-Artifacts-For-Apireview($ArtifactPath, $PackageName)
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:58]: Where is this method? `Find-Artifacts-For-Apireview`

File: eng/common/scripts/Create-APIReview.ps1 [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  [Parameter(Mandatory=$True)]
+  [string] $APIKey,
+  [Parameter(Mandatory=$True)]
+  [string] $APILabel,
+  [string] $PackageName = ""
+)
+
+
+# Submit API review request and return status whether current revision is approved or pending or failed to create review
+function Submit-APIReview($packagename, $filePath, $uri, $apiKey, $apiLabel){
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:16]: nit: I'd put the `{` on its own line for larger blocks like a function.

File: eng/common/scripts/Create-APIReview.ps1 [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    return $StatusCode
+}
+
+
+. "${PSScriptRoot}\..\common\scripts\common.ps1"
+$packages = Find-Artifacts-For-Apireview($ArtifactPath, $PackageName)
+$responses = @{}
+if ($packages){
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:60]: nit: If you do keep `{` on the same line there should be a space in front of it.

File: eng/common/scripts/Create-APIReview.ps1 [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  [Parameter(Mandatory=$True)]
+  [string] $APILabel,
+  [string] $PackageName = ""
+)
+
+
+# Submit API review request and return status whether current revision is approved or pending or failed to create review
+function Submit-APIReview($packagename, $filePath, $uri, $apiKey, $apiLabel){
+    
+    $multipartContent = [System.Net.Http.MultipartFormDataContent]::new()
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:18]: Is calling all the .NET APIs here the best way to do this in PS? If this is the only way then so be it but this definitely looks a little unusual for PS code.

File: eng/common/scripts/Create-APIReview.ps1 [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    $stringHeader = [System.Net.Http.Headers.ContentDispositionHeaderValue]::new("form-data")
+    $stringHeader.Name = "label"
+    $StringContent = [System.Net.Http.StringContent]::new($apiLabel)
+    $StringContent.Headers.ContentDisposition = $stringHeader
+    $multipartContent.Add($stringContent)
+
+    $form = @{
+        'label' = $apiLabel;
+        'file' = Get-content -Path $filePath
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:37]: We might want to use `-Raw` for the Get-Content call.

File: eng/common/scripts/Create-APIReview.ps1 [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+if ($packages){
+    foreach($pkg in $packages.GetEnumerator()){
+        Write-Host "Submitting API Review for package $($pkg.Key)"
+        Write-Host $packages[$pkg.Key]
+        $responses[$pkg.Key] = Submit-APIReview -packagename $pkg.Key -filePath $packages[$pkg.Key] -uri $APIViewUri -apiKey $APIKey -apiLabel $APILabel
+    }
+}
+
+$FoundFailure = $False
+foreach($res in $responses.GetEnumerator()){
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:69]:
```suggestion
foreach($pkgName in $responses.Keys)){
```

File: eng/common/scripts/Create-APIReview.ps1 (hunk 1 above)
Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1:58]: This is in each Language-Settings.ps1 to have individual filtering conditions as well as language specific instructions to generate ast files.

File: eng/common/scripts/Create-APIReview.ps1 (hunk 4 above)
Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1:18]: Basic http request in powershell is not binding uploaded file to form fields in asp.net app ( IFormFile). This form has mixed form data content so I had to use .Net modules. Alternate choice was to use python script to keep it simple but I thought of going with PS script.

File: eng/common/scripts/Create-APIReview.ps1 (hunk 2 above)
Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1]: Fixed

File: eng/common/scripts/Create-APIReview.ps1 (hunk 5 above)
Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1]: This object is not used. I've removed this code.

File: eng/common/scripts/Create-APIReview.ps1 (hunk 6 above)
Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1]: Updated

File: eng/common/scripts/Create-APIReview.ps1 (hunk 1 above)
Comment (weshaggard) [reply eng/common/scripts/Create-APIReview.ps1:55]: We should check to see that it exists before calling it.  Something like https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/artifact-metadata-parsing.ps1#L96.

Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1]: Fixed

**Comment by praveenkuttappan** [comment]
@weshaggard Please let me know if any additional change is required. I have changes for .Net and Java but waiting to get this merged.

File: eng/common/scripts/Create-APIReview.ps1 [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    return $StatusCode
+}
+
+
+. (Join-Path $PSScriptRoot common.ps1)
+$packages = @{}
+
+if (Test-Path "Function:Find-Artifacts-For-Apireview")
+{
+    $packages = Find-Artifacts-For-Apireview($ArtifactPath, $PackageName)
```
Comment (weshaggard) [inline eng/common/scripts/Create-APIReview.ps1:59]:
```suggestion
    $packages = Find-Artifacts-For-Apireview $ArtifactPath $PackageName
```
You cannot call this as a standard function in PS the are more like command line arguments. Have you ran this code yet? If not I'd like to wait until we can test it in the repo's before we merge this into eng\common.

**Review by weshaggard — APPROVED** [review]
Left one more comment which I think needs to be addressed otherwise I don't believe the script will actually work. Please also be sure to test this in at least one of the language repo's before merging this into the tools repo.

File: eng/common/scripts/Create-APIReview.ps1 (hunk 7 above)
Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1:59]: This did work actually and my test run had created below API review. https://apiviewstaging.azurewebsites.net/Assemblies/Review/609cf46560804ae6858e9aab786e73e5

I will anyway change to code to PS standard format as per suggestion above.

Comment (praveenkuttappan) [reply eng/common/scripts/Create-APIReview.ps1]: Created a test review for Java using latest changes: https://apiviewstaging.azurewebsites.net/Assemblies/Review/d2cfa2fb40de477491ee614f826e83fc

**Pull request closed without merging**